tests/VatServiceTest.php

```diff
@@ -31,4 +31,31 @@ public function countryTestData(): Generator
         yield [false, true, false];
         yield [false, false, false];
     }
+
+    /** @dataProvider euVatRateTestData */
+    public function testEuropeanVatRate(string $countryCode, bool $validCountry, bool $inEu, ?float $rate, float $result): void
+    {
+        $isoMock = $this->createMock(Iso2::class);
+        $isoMock->method('isCountryValid')->willReturn($validCountry);
+        $isoMock->method('isCountryInEu')->willReturn($inEu);
+
+        $vatResolverMock = $this->createMock(ResolvesVatRates::class);
+        $vatResolverMock->method('getDefaultVatRateForCountry')->willReturn($rate);
+
+        $service = new VatStub();
+        $service->setCountries($isoMock);
+        $service->setVatRateResolver($vatResolverMock);
+
+        Assert::assertEquals($result, $service->europeanVatRate($countryCode));
```

Comment:
`assertSame` is safer than `assertEquals`. same does `===`, and equals does `==`

Reply:
Oh my, I thought it was the other way around.. Now I have to refactor some code in other projects I'm afraid 😂

Reply:
I have just pushed changes updating this. :)